File: app/src/main/java/de/bikebean/app/ui/drawer/history/position/PositionHistoryAdapter.java
package de.bikebean.app.ui.drawer.history.position;

import android.content.Context;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TableLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.cardview.widget.CardView;

import java.util.List;
import java.util.Locale;

import de.bikebean.app.MainActivity;
import de.bikebean.app.R;
import de.bikebean.app.db.DatabaseEntity;
import de.bikebean.app.db.state.LocationState;
import de.bikebean.app.ui.drawer.history.HistoryAdapter;
import de.bikebean.app.ui.utils.date.DateUtils;

class PositionHistoryAdapter extends HistoryAdapter {

    static class PositionHistoryViewHolder extends HistoryViewHolder {
        private final TableLayout table;
        private final CardView buttonOpenMap;
        private final TextView lat, lng, acc, dateTimeText, smsIdText;

        PositionHistoryViewHolder(final @NonNull View v) {
            super(v);

            table = v.findViewById(R.id.tableLayout2);
            buttonOpenMap = v.findViewById(R.id.buttonOpenMap);
            lat = v.findViewById(R.id.lat2);
            lng = v.findViewById(R.id.lng2);
            acc = v.findViewById(R.id.acc2);
            dateTimeText = v.findViewById(R.id.dateTimeText6);
            smsIdText = v.findViewById(R.id.smsIdText);
        }
    }

    PositionHistoryAdapter(final @NonNull Context context,
                           final @Nullable List<? extends DatabaseEntity> states) {
        super(context, states);
    }

    @NonNull
    @Override
    public PositionHistoryViewHolder onCreateViewHolder(final @NonNull ViewGroup parent, int viewType) {
        final @NonNull View itemView = mInflater.inflate(
                R.layout.recyclerview_item_position_history,
                parent, false
        );
        return new PositionHistoryViewHolder(itemView);
    }

    @Override
    public void onBindViewHolder(final @NonNull HistoryViewHolder holder, int position) {
        final @NonNull PositionHistoryViewHolder pHolder = (PositionHistoryViewHolder) holder;

        if (mStates != null) {
            final @NonNull LocationState current = (LocationState) mStates.get(position);

            pHolder.table.setVisibility(View.VISIBLE);
            pHolder.lat.setText(String.format(Locale.GERMANY, "%.7f", current.getLat()));
            pHolder.lng.setText(String.format(Locale.GERMANY, "%.7f", current.getLng()));
            pHolder.acc.setText(String.format(Locale.GERMANY, "%.1f", current.getAcc()));
            pHolder.dateTimeText.setText(DateUtils.convertPeriodToHuman(current.getTimestamp()));
            pHolder.smsIdText.setText(
                    String.format(Locale.GERMANY, "SmsId: %d", current.getSmsId())
            );

            pHolder.buttonOpenMap.setOnClickListener(v ->
                ((MainActivity) ctx).navigateTo(R.id.map_action, current.getArgs())
            );
        } else {
            pHolder.table.setVisibility(View.INVISIBLE);
            pHolder.buttonOpenMap.setVisibility(View.INVISIBLE);
            pHolder.dateTimeText.setVisibility(View.GONE);
            pHolder.smsIdText.setVisibility(View.GONE);
        }
    }
}

File: app/src/main/java/de/bikebean/app/ui/drawer/status/location/history/PositionHistoryAdapter.kt
package de.bikebean.app.ui.drawer.status.location.history

import android.content.Context
import android.view.View
import android.view.ViewGroup
import android.widget.TableLayout
import android.widget.TextView
import androidx.cardview.widget.CardView
import de.bikebean.app.MainActivity
import de.bikebean.app.R
import de.bikebean.app.db.DatabaseEntity
import de.bikebean.app.db.state.LocationState
import de.bikebean.app.ui.drawer.status.history.HistoryAdapter
import de.bikebean.app.ui.utils.date.DateUtils.convertPeriodToHuman

internal class PositionHistoryAdapter(
        context: Context,
        states: List<DatabaseEntity?>?
) : HistoryAdapter(context, states) {

    internal class PositionHistoryViewHolder(v: View) : HistoryViewHolder(v) {
        val table: TableLayout = v.findViewById(R.id.tableLayout2)
        val buttonOpenMap: CardView = v.findViewById(R.id.buttonOpenMap)
        val lat: TextView = v.findViewById(R.id.lat2)
        val lng: TextView = v.findViewById(R.id.lng2)
        val acc: TextView = v.findViewById(R.id.acc2)
        val dateTimeText: TextView = v.findViewById(R.id.dateTimeText6)
        val smsIdText: TextView = v.findViewById(R.id.smsIdText)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): PositionHistoryViewHolder =
            PositionHistoryViewHolder(mInflater.inflate(
                    R.layout.recyclerview_item_position_history,
                    parent, false
            ))

    override fun onBindViewHolder(holder: HistoryViewHolder, position: Int) {
        mStates?.let {
            (it[position] as LocationState).let { state ->
                (holder as PositionHistoryViewHolder).apply {
                    table.visibility = View.VISIBLE
                    lat.text = String.format("%.7f", state.lat)
                    lng.text = String.format("%.7f", state.lng)
                    acc.text = String.format("%.1f", state.acc)
                    dateTimeText.text = convertPeriodToHuman(state.timestamp)
                    smsIdText.text = String.format("SmsId: %d", state.smsId)
                    buttonOpenMap.setOnClickListener {
                        (ctx as MainActivity).navigateTo(R.id.map_action, state.args)
                    }
                }
            }
        } ?: run {
            (holder as PositionHistoryViewHolder).apply {
                table.visibility = View.INVISIBLE
                buttonOpenMap.visibility = View.INVISIBLE
                dateTimeText.visibility = View.GONE
                smsIdText.visibility = View.GONE
            }
        }
    }
}